run.c

```diff
@@ -57,8 +57,8 @@ void run(void)
 
     write_cpu_log(NumCurrentTiStep); /* produce some CPU usage info */
 
-    /* find the first output time, -1 is to allow immediate output at first step. */
-    inttime_t Ti_nextoutput = find_next_outputtime(All.Ti_Current - 1);
+    /* find the first output time. */
+    inttime_t Ti_nextoutput = find_next_outputtime(All.Ti_Current);
```

Comment:
I have doubts about this change. Without the -1 even if the first redshift is in outputlist no output on first step is made.

Reply:
But with it output is always made on the first step, which is probably not what we want. Why do we want output on the first step? Would an extra RestartFlag serve the purpose?

Reply:
If a snapshot is requested on the first step, then the code shall produce a snapshot. It is the least surprising behavior. Practically, a snapshot on the first step will run through all modules in a realistic setting; so it will crash early if there are mis-configurations.

Reply:
I agree it is a bug that when the first redshift is not requested a snapshot is always written, but the issue is probably in find_next_outputtime, not in this -1.

Reply:
I see that the initial time is added to OutputTimeList in init.c to avoid special conditioning between the starting point to the end point. It is not an OutputTimeList item though. To fix this I think we need to introduce a new concept, 'SyncPoints', which contains all OutputTimes. The code will do PM steps on SyncPoints for sure, and we will use the last syncpoint prior to outputtime (up to a small tolerance) to dump the snapshot.

Reply:
I'll give this a go. I think these sync points can serve a way to avoid running out of timebins when snapshots are too sparse, too.

Reply:
Take a look at this https://github.com/rainwoodman/MP-Gadget/pull/109

Reply:
Looks a good idea.